### PR TITLE
feat: implement secrets handling

### PR DESCRIPTION
**Pull request by sgaist**
## Describe your changes

This PR implements the support for user secrets handling.

This is twofold:

- The web API now accepts a new field which provides a list of secret ids and the path where they should be made available
- The patches will now contain a new init container and volumes to retrieve secrets content and decrypt them to make them available to the main container

## Issue ticket number and link

Part of #1206
Part of https://github.com/SwissDataScienceCenter/renku-design-docs/pull/41

/deploy #notest renku-ui=lorenzo/shape-up-secrets-in-sessions renku=build/secrets-in-sessions renku-gateway=limit-user-secret renku-data-services=pitch/secret-storage extra-values=secretsStorage.encryptionKey=vMqdVXcSIwjwONvGiNE2vjzRjmi8NJ8Y9lTB8m3l/Xg=

### CODE REVIEW

**Comment by rokroskar** [comment]
@Panaetius presumably this PR can be deployed with the other components to test it live?

**Review by Panaetius — COMMENTED** [review]
This looks really nice, thanks a lot!

File: renku_notebooks/api/notebooks.py [hunk 1]
```diff
@@ -460,6 +473,38 @@ def launch_notebook_helper(
...
+        }
+
+        def _on_error(error_msg):
+            config.k8s.client.delete_server(
+                server.server_name, forced=True, safe_username=user.safe_username
+            )
+            raise RuntimeError(error_msg)
+
+        try:
+            answer = requests.post(
```
[the remark this answers is not in the document]
Comment (sgaist) [reply renku_notebooks/api/notebooks.py:498]: Good point, fixed !

File: renku_notebooks/api/schemas/secrets.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    def _deserialize(self, value, attr, data, **kwargs):
+        path = Path(value)
+        if not path.is_absolute():
+            raise ValidationError("Path is not aboslute")
+        return str(path)
+
+
+class UserSecrets(Schema):
+    # List of ids of the user's secrets
+    user_secret_ids = fields.List(fields.Str, required=True)
```
[the remark this answers is not in the document]
Comment (sgaist) [reply renku_notebooks/api/schemas/secrets.py]: Interesting !

Done

**Comment by Panaetius** [comment]
This confuses me quite a bit, the deployment fails due to this error in the Notebooks: 
```
[2024-05-02 07:31:01 +0000] [8] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/app/.venv/lib/python3.11/site-packages/gunicorn/arbiter.py", line 609, in spawn_worker
    worker.init_process()
  File "/app/.venv/lib/python3.11/site-packages/gunicorn/workers/ggevent.py", line 147, in init_process
    super().init_process()
  File "/app/.venv/lib/python3.11/site-packages/gunicorn/workers/base.py", line 134, in init_process
    self.load_wsgi()
  File "/app/.venv/lib/python3.11/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
    self.wsgi = self.app.wsgi()
                ^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.11/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
                    ^^^^^^^^^^^
  File "/app/.venv/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
    return self.load_wsgiapp()
           ^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
    return util.import_app(self.app_uri)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.11/site-packages/gunicorn/util.py", line 371, in import_app
    mod = importlib.import_module(module)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1126, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/app/renku_notebooks/__init__.py", line 27, in <module>
    from .api.notebooks import (
  File "/app/renku_notebooks/api/notebooks.py", line 45, in <module>
    from .classes.server import Renku2UserServer, Repository, UserServer
  File "/app/renku_notebooks/api/classes/server.py", line 23, in <module>
    from ..schemas.secrets import K8sUserSecrets
  File "/app/renku_notebooks/api/schemas/secrets.py", line 4, in <module>
    from ulid import ULID
ModuleNotFoundError: No module named 'ulid'
```
but I can see that ulid was added in pyproject.toml . Not really sure what is going on there

**Comment by sgaist** [comment]
> This confuses me quite a bit, the deployment fails due to this error in the Notebooks:
> 
> SNIPPED
> 
> but I can see that ulid was added in pyproject.toml . Not really sure what is going on there

I am as lost as you are, I am wondering if there's some caching issue at play.

**Comment by Panaetius** [comment]
> > This confuses me quite a bit, the deployment fails due to this error in the Notebooks:
> > SNIPPED
> > but I can see that ulid was added in pyproject.toml . Not really sure what is going on there
> 
> I am as lost as you are, I am wondering if there's some caching issue at play.

the image contains the git sha as part of the tag, and that is correct, so I don't think it could be cache related. I'll look into it more later today

**Pull request merged**